### PR TITLE
[#1667] Refactor AnnotatorAnalyzer class

### LINKED ISSUE #1667: Refactoring `AnnotatorAnalyzer` class
**What ideas would you like to see in RepoSense?**

Refactoring the `AnnotatorAnalyzer` class.

**Is the feature request related to a problem?**

The current `AnnotatorAnalyzer` class' methods are difficult to understand. There are many null checks even though the optional class is being used. Additionally, there is a `ArrayIndexOutOfBoundsException` caught that seems to perform normal program flow logic. 

https://github.com/reposense/RepoSense/blob/ed131f6294bf83fac436a6265bc4d53b022739f7/src/main/java/reposense/authorship/analyzer/AnnotatorAnalyzer.java#L99-L103

The lines above are not tested by any test cases as well.

**If possible, describe the solution**

Change the logic in `extractAuthorName` to prevent the index exception from occurring at all.
Additionally, make use of `Optional` class methods like `filter` and `map` to perform operations instead of checking against `null`. 

**If applicable, describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by chan-j-d**
Fixes #1667 

Key implementation changes: 
- `try-catch` block has been removed. The introduction of it was from this commit https://github.com/reposense/RepoSense/commit/ab6ec641b43ec59222a6632b9c6b8201560ab5cc where it makes sense. However, in the current version, there does not seem to be a way this exception gets thrown as the only array access is performed in `extractAuthorName` which already performs length check before accessing via index. 
- On encountering an author tag without author name specified, `findAuthorInLine` performs the logic of deciding if this is an end of annotated segment or the start of an annotated segment by an unknown author. That logic has been moved to `aggregateAnnotationAuthorInfo`. `findAuthorInLine` now only strictly returns the `Author` object corresponding to what is found in the line. 

Commit message:
```
The AnnotatorAnalyzer class is difficult to understand. It also performs
many null checks even while the Optional class is used. Exception
handling is also used in one part of the normal program flow.

Let's refactor the methods to improve the readability and quality.
```

### CODE REVIEW

**Review by gok99 — APPROVED** [review]
LGTM! This is vastly more readable, nice work :)

File: src/main/java/reposense/authorship/analyzer/AnnotatorAnalyzer.java [hunk 1]
```diff
@@ -45,98 +45,82 @@
...
 
     /**
-     * Checks if the line is a valid @@author tag comment line
+     * Returns the index representing the type of comment the @@author tag line is.
+     *
      * @param line The line to be checked
      * @return The index of the comment if the comment pattern matches, -1 if no match could be found
      */
-    public static int checkValidCommentLine(String line) {
+    public static int getCommentType(String line) {
```
[the remark this answers is not in the document]
Comment (chan-j-d) [reply src/main/java/reposense/authorship/analyzer/AnnotatorAnalyzer.java:123]: Hmm maybe something like `getCommentTypeIndex`? I think `CommentIndex` alone doesn't quite describe what it is. Basically, there is a list of hardcoded supported comment syntax types in this class for different languages. The index returned here is the index of that comment syntax type within the list.

[the remark this answers is not in the document]
Comment (Zhou-Jiahao-1998) [reply src/main/java/reposense/authorship/analyzer/AnnotatorAnalyzer.java]: Sure. I think that is better.

**Review by Zhou-Jiahao-1998 — APPROVED** [review]
LGTM

**Review by dcshzj — APPROVED** [review]
LGTM! Excellent work on this refactoring, thank you!

**Pull request merged**